include/knowhere/sparse_utils.h

```diff
@@ -0,0 +1,226 @@
+        return !(lhs < rhs);
+    }
+
+    inline friend bool
+    operator==(const IdVal& lhs, const IdVal& rhs) {
+        return lhs.id == rhs.id && lhs.val == rhs.val;
+    }
+};
+
+template <typename T>
+class SparseRow {
+    static_assert(std::is_same_v<T, fp32>, "SparseRow supports float only");
+
+ public:
+    SparseRow() = default;
+
+    SparseRow(const SparseRow& other) {
+        copy_from(other);
+    }
+
+    SparseRow&
+    operator=(const SparseRow& other) {
+        copy_from(other);
+        return *this;
+    }
+
+    ~SparseRow() {
+        if (own_data && data != nullptr) {
+            delete[] data;
+        }
+    }
+
+    size_t
+    size() const {
+        return count;
+    }
+
+    size_t
+    byte_size() const {
+        return count * (sizeof(table_t) + sizeof(T)) + sizeof(*this);
+    }
+
+    void*
+    data_ptr() const {
+        return data;
+    }
+
+    // dim of a sparse vector is the max index + 1, or 0 for an empty vector.
+    int64_t
+    dim() const {
+        if (count == 0)
+            return 0;
+        return (*this)[count - 1].id + 1;
+    }
+
+    IdVal<T>
+    operator[](size_t i) const {
+        table_t index = *(table_t*)(data + i * (sizeof(table_t) + sizeof(T)));
+        T value = *(T*)(data + i * (sizeof(table_t) + sizeof(T)) + sizeof(table_t));
+        return {index, value};
+    }
+
+    void
+    set_at(size_t i, table_t index, T value) {
+        *(table_t*)(data + i * (sizeof(table_t) + sizeof(T))) = index;
+        *(T*)(data + i * (sizeof(table_t) + sizeof(T)) + sizeof(table_t)) = value;
+    }
+
+    float
+    dot(const SparseRow<T>& other) const {
+        float product_sum = 0.0f;
+        size_t i = 0, j = 0;
+        while (i < count && j < other.count) {
+            auto [index1, value1] = (*this)[i];
+            auto [index2, value2] = other[j];
+
+            if (index1 < index2) {
+                ++i;
+            } else if (index1 > index2) {
+                ++j;
+            } else {
+                product_sum += value1 * value2;
+                ++i;
+                ++j;
+            }
+        }
+        return product_sum;
+    }
+
+    void
+    reserve(size_t count) {
+        if (own_data && data != nullptr) {
+            delete[] data;
+        }
+        data = new uint8_t[count * (sizeof(table_t) + sizeof(T))];
+        own_data = true;
+        this->count = count;
+    }
+
+    // performs deep copy of data
+    void
+    copy_from(const SparseRow<T>& other) {
+        reserve(other.count);
+        memcpy(data, other.data, count * (sizeof(table_t) + sizeof(T)));
+    }
+
+ private:
+    // data must be sorted by column id. use raw pointer for easy mmap and zero
+    // copy.
+    uint8_t* data = nullptr;
+    size_t count = 0;
+    bool own_data = false;
+};
+
+// When pushing new elements into a MaxMinHeap, only `capacity` elements with the
+// largest val are kept. pop()/top() returns the smallest element out of them.
+template <typename T>
```

Comment:
Don't we have any similar structures in our code already? I'm just curious, the implementation is fine